src/main/java/in/zapr/druid/druidry/client/DruidJerseyClient.java

```diff
@@ -0,0 +1,125 @@
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.GenericType;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import in.zapr.druid.druidry.client.exception.ConnectionException;
+import in.zapr.druid.druidry.client.exception.QueryException;
+import in.zapr.druid.druidry.query.DruidQuery;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class DruidJerseyClient implements DruidClient {
+
+    private static final int DEFAULT_CONNECTION_POOL_LIMIT = 5;
+    private final String druidUrl;
+    private final DruidConfiguration druidConfiguration;
+
+    private ClientConfig jerseyConfig;
+    private Client client;
+    private WebTarget queryWebTarget;
+
+    public DruidJerseyClient(@NonNull DruidConfiguration druidConfiguration,
+                             ClientConfig jerseyConfig) {
+
+        this.druidUrl = druidConfiguration.getUrl();
+        this.jerseyConfig = jerseyConfig;
+
+        log.debug("Will query to {}", druidUrl);
+        this.druidConfiguration = druidConfiguration;
+    }
+
+    @Override
+    public void connect() throws ConnectionException {
+
+        PoolingHttpClientConnectionManager connectionManager
+                = new PoolingHttpClientConnectionManager();
+
+        int numberOfConnectionsInPool = DEFAULT_CONNECTION_POOL_LIMIT;
+
+        if (this.druidConfiguration.getConcurrentConnectionsRequired() != null) {
+            numberOfConnectionsInPool = this.druidConfiguration.getConcurrentConnectionsRequired();
+        }
+
+        connectionManager.setDefaultMaxPerRoute(numberOfConnectionsInPool);
+
+        try {
+            if (jerseyConfig == null) {
+                this.jerseyConfig = new ClientConfig();
+            }
+
+            this.jerseyConfig.property(ApacheClientProperties.CONNECTION_MANAGER, connectionManager);
+            this.jerseyConfig.connectorProvider(new ApacheConnectorProvider());
+
+            this.client = ClientBuilder.newClient(this.jerseyConfig);
+            this.queryWebTarget = this.client.target(this.druidUrl);
+
+        } catch (Exception e) {
+            throw new ConnectionException(e);
+        }
+    }
+
+    @Override
+    public void close() throws ConnectionException {
+        try {
+            this.client.close();
+        } catch (Exception e) {
+            throw new ConnectionException(e);
+        }
+    }
+
+    @Override
+    public String query(DruidQuery druidQuery) throws QueryException {
+
+        try (Response response = this.queryWebTarget
+                .request(MediaType.APPLICATION_JSON)
+                .post(Entity.entity(druidQuery, MediaType.APPLICATION_JSON))) {
+
+            if (response.getStatus() == Response.Status.INTERNAL_SERVER_ERROR.getStatusCode()) {
+                handleInternalServerResponse(response);
+            }
+
+            return response.readEntity(String.class);
+
+        } catch (Exception e) {
```

Comment:
You need to add a catch block for QueryException as well which could be thrown from handleInternalServerResponse(). You could return the same QueryException back to the caller